File: 11/gemini-3.0-pro/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

var graph = make(map[string][]string)
var memo1 = make(map[string]int64)
var memo2 = make(map[string]int64)

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if strings.TrimSpace(line) == "" {
			continue
		}
		parts := strings.Split(line, ": ")
		source := parts[0]
		targets := strings.Split(parts[1], " ")
		graph[source] = targets
	}

	paths1 := countPathsP1("you")
	fmt.Printf("Day 11 Part 1: %d\n", paths1)

	paths2 := countPathsP2("svr", 0)
	fmt.Printf("Day 11 Part 2: %d\n", paths2)
}

func countPathsP1(current string) int64 {
	if current == "out" {
		return 1
	}
	if val, ok := memo1[current]; ok {
		return val
	}

	if _, ok := graph[current]; !ok {
		return 0
	}

	var total int64 = 0
	for _, neighbor := range graph[current] {
		total += countPathsP1(neighbor)
	}

	memo1[current] = total
	return total
}

func countPathsP2(current string, mask int) int64 {
	if current == "dac" {
		mask |= 1
	} else if current == "fft" {
		mask |= 2
	}

	if current == "out" {
		if mask == 3 {
			return 1
		}
		return 0
	}

	key := fmt.Sprintf("%s-%d", current, mask)
	if val, ok := memo2[key]; ok {
		return val
	}

	if _, ok := graph[current]; !ok {
		return 0
	}

	var total int64 = 0
	for _, neighbor := range graph[current] {
		total += countPathsP2(neighbor, mask)
	}

	memo2[key] = total
	return total
}

File: 11/gemini-3.0-pro/rust/src/main.rs
use std::collections::HashMap;
use std::fs;

fn main() {
    let input = fs::read_to_string("input.txt").expect("Unable to read file");
    let mut graph: HashMap<String, Vec<String>> = HashMap::new();

    for line in input.lines() {
        if line.trim().is_empty() {
            continue;
        }
        let parts: Vec<&str> = line.split(": ").collect();
        let source = parts[0].to_string();
        let targets: Vec<String> = parts[1].split_whitespace().map(|s| s.to_string()).collect();
        graph.insert(source, targets);
    }

    // Part 1
    let mut memo1: HashMap<String, u64> = HashMap::new();
    let paths1 = count_paths_p1("you", &graph, &mut memo1);
    println!("Day 11 Part 1: {}", paths1);

    // Part 2
    let mut memo2: HashMap<(String, u8), u64> = HashMap::new();
    let paths2 = count_paths_p2("svr", 0, &graph, &mut memo2);
    println!("Day 11 Part 2: {}", paths2);
}

fn count_paths_p1(current: &str, graph: &HashMap<String, Vec<String>>, memo: &mut HashMap<String, u64>) -> u64 {
    if current == "out" {
        return 1;
    }
    if let Some(&count) = memo.get(current) {
        return count;
    }

    let mut total = 0;
    if let Some(neighbors) = graph.get(current) {
        for neighbor in neighbors {
            total += count_paths_p1(neighbor, graph, memo);
        }
    }

    memo.insert(current.to_string(), total);
    total
}

fn count_paths_p2(current: &str, mut mask: u8, graph: &HashMap<String, Vec<String>>, memo: &mut HashMap<(String, u8), u64>) -> u64 {
    if current == "dac" {
        mask |= 1;
    } else if current == "fft" {
        mask |= 2;
    }

    if current == "out" {
        return if mask == 3 { 1 } else { 0 };
    }

    let key = (current.to_string(), mask);
    if let Some(&count) = memo.get(&key) {
        return count;
    }

    let mut total = 0;
    if let Some(neighbors) = graph.get(current) {
        for neighbor in neighbors {
            total += count_paths_p2(neighbor, mask, graph, memo);
        }
    }

    memo.insert(key, total);
    total
}
